### PR TITLE
Add GPU pooling kernel

### PR DESCRIPTION
**Pull request by pfultz2**

### CODE REVIEW

**Review by CharlieL7 — COMMENTED** [review]
Not getting rid of the miopen pooling `src/targets/gpu/pooling.cpp`?

**Comment by pfultz2** [comment]
> Not getting rid of the miopen pooling src/targets/gpu/pooling.cpp?

Not yet, good to have it as a fallback just in case we find a bug.

File: test/verify/test_lpnorm_pooling_pad.cpp [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    {
+        migraphx::program p;
+        auto* mm = p.get_main_module();
+        auto x   = mm->add_parameter("x", {migraphx::shape::float_type, {1, 3, 5, 5}});
+        mm->add_instruction(migraphx::make_op("pooling",
+                                              {{"mode", migraphx::op::pooling_mode::lpnorm},
+                                               {"padding", {2, 2}},
+                                               {"stride", {1, 1}},
+                                               {"lengths", {3, 3}},
+                                               {"lp_order", 2}}),
```
Comment (umangyadav) [inline test/verify/test_lpnorm_pooling_pad.cpp:42]: What's the difference between this test and `test_lpnorm_pooling` ? i do not find any diff. Can you do count_include_pad test for the maxpool and lpnorm ?

Comment (umangyadav) [reply test/verify/test_lpnorm_pooling_pad.cpp:42]: add a few tests with `ceil_mode` enabled

File: src/targets/gpu/kernels/include/migraphx/kernels/pooling.hpp [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+    MIGRAPHX_DEVICE_CONSTEXPR T final(T x, index_int) const
+    {
+        if constexpr(P == 0)
+            return 1;
+        else if constexpr(P == 1)
+            return x;
+        else if constexpr(P == 2)
+            return migraphx::sqrt(x);
+        else
+            return migraphx::pow(x, 1. / P);
```
Comment (umangyadav) [inline src/targets/gpu/kernels/include/migraphx/kernels/pooling.hpp:107]: `1/P` would it be considered as constexpr ?

Comment (pfultz2) [reply src/targets/gpu/kernels/include/migraphx/kernels/pooling.hpp:107]: What do you mean?

Comment (pfultz2) [reply src/targets/gpu/kernels/include/migraphx/kernels/pooling.hpp:107]: Ah you mean can it be const-folded?

File: test/verify/test_lpnorm_pooling_pad.cpp (hunk 1 above)
Comment (pfultz2) [reply test/verify/test_lpnorm_pooling_pad.cpp:42]: This is only supposed to have padding, let me update the other tests to remove the padding.

File: src/targets/gpu/kernels/include/migraphx/kernels/pooling.hpp (hunk 2 above)
Comment (umangyadav) [reply src/targets/gpu/kernels/include/migraphx/kernels/pooling.hpp:107]: Yes i mean if it can be const folded.

**Comment by pfultz2** [comment]
> From what I can figure out, it looks like each thread calculates a window and then moves on to another window depending on the number of threads.

Its a lane reduction, so one workitem does one window for pooling, this is similiar to miopen.

> Have we considered other work distribution schemes?

Yes we can do a  "blockwise"(or it might be smaller than a block or wave) reduction across the window and do a lane reduction down the window. This should be faster, but I plan to implement this as enhancement in the future.

File: src/targets/gpu/kernels/include/migraphx/kernels/pooling.hpp (hunk 2 above)
Comment (pfultz2) [reply src/targets/gpu/kernels/include/migraphx/kernels/pooling.hpp:107]: Yea it should be const folded by the compiler.

**Review by CharlieL7 — APPROVED** [review]
LGTM, do we have an issue tracking the "block-wise reduction" enhancement?

File: test/verify/test_lpnorm_pooling_pad.cpp (hunk 1 above)
Comment (pfultz2) [reply test/verify/test_lpnorm_pooling_pad.cpp:42]: Add `ceil_mode` tests for lpnorm as well.

**Comment by pfultz2** [comment]
> LGTM, do we have an issue tracking the "block-wise reduction" enhancement?

https://github.com/ROCm/AMDMIGraphX/issues/3175

File: src/targets/cpu/lowering.cpp [hunk 3]
```diff
@@ -415,12 +415,23 @@ struct cpu_apply
                        {ins->inputs().front()});
     }
 
+    // TODO:  update lowering to run the reference
+    // code when OneDNN can't execute pooling for a CPU
```
Comment (umangyadav) [inline src/targets/cpu/lowering.cpp:419]: Lowering to reference backend can be a good first issue.

**Review by umangyadav — APPROVED** [review]

File: src/include/migraphx/op/pooling.hpp [hunk 4]
```diff
@@ -344,9 +331,8 @@ struct pooling
                 int end;
                 std::size_t dilated_kernel_dim = dilate_dim(kernel_dims[d_2], dilations[d_2]);
                 // NOLINT
-                if(count_include_pad and ceil_mode and (mode != pooling_mode::max))
+                if(count_include_pad and (mode != pooling_mode::max))
```
Comment (umangyadav) [inline src/include/migraphx/op/pooling.hpp:334]: Comments on line 346 
"In non ceiling mode" 
requires update

**Review by umangyadav — CHANGES_REQUESTED** [review]

**Review by umangyadav — APPROVED** [review]

**Comment by lakhinderwalia** [comment]
This PR breaks Resnet Model accuracy tests. 
See: https://ontrack-internal.amd.com/browse/SWDEV-471087

**Pull request merged**